src/LinearOperators.jl

```diff
@@ -85,346 +90,355 @@ end
+  LinearOperator{T}(size(M)..., symmetric, hermitian,
+                    v -> M * v,
+                    Nullable{Function}(u -> M.' * u),
+                    Nullable{Function}(w -> M' * w))
 
 """Constructs a linear operator from a symmetric tridiagonal matrix. If
 its elements are real, it is also Hermitian, otherwise complex
 symmetric."""
-LinearOperator(M :: SymTridiagonal) =
-    LinearOperator(M; symmetric = true, hermitian = eltype(M) <: Real)
+LinearOperator{T}(M :: SymTridiagonal{T}) =
+    LinearOperator(M; symmetric=true, hermitian=eltype(M) <: Real)
 
+# the only advantage of this constructor is optional args
+# use LinearOperator{Float64} if you mean real instead of complex
 "Construct a linear operator from functions."
-LinearOperator(nrow :: Int, ncol :: Int, dtype :: DataType,
-               symmetric :: Bool, hermitian :: Bool,
-               prod :: Function, tprod :: Function, ctprod :: Function) =
-  LinearOperator(nrow, ncol, dtype, symmetric, hermitian,
-                 prod, Nullable{Function}(tprod), Nullable{Function}(ctprod))
-
-"Construct a real symmetric linear operator from a function."
-LinearOperator(nrow :: Int, dtype :: DataType, prod :: Function) =
-  LinearOperator(nrow, nrow, dtype, true, true,
-                 prod,
-                 Nullable{Function}(prod),
-                 Nullable{Function}(prod))
-
-"Construct a linear operator from a single function."
-LinearOperator(nrow :: Int, ncol :: Int, dtype :: DataType,
-               symmetric :: Bool, hermitian :: Bool,
-               prod :: Function) =
-  LinearOperator(nrow, ncol, dtype, symmetric, hermitian,
-                 prod, Nullable{Function}(), Nullable{Function}())
-
-
-if VERSION ≥ v"0.4.0-dev"
-  import Base.+, Base.-, Base.*, Base.(.+), Base.(.-), Base.(.*)
-  import Base.transpose, Base.ctranspose
+function LinearOperator(nrow :: Int, ncol :: Int,
+                        symmetric :: Bool, hermitian :: Bool,
+                        prod :: Function,
+                        tprod :: Union{Function,Nullable{Function}}=Nullable{Function}(),
+                        ctprod :: Union{Function,Nullable{Function}}=Nullable{Function}())
+
+  T = hermitian ? (symmetric ? Float64 : Complex128) : Complex128
+  LinearOperator{T}(nrow, ncol, symmetric, hermitian, prod, tprod, ctprod)
 end
 
+# "Construct a real symmetric or complex hermitian linear operator from a function."
+# LinearOperator(nrow :: Int, T :: DataType, prod :: Function) =
+#   LinearOperator{T}(nrow, nrow, T <: Real, T <: Real || T <: Complex,
+#                     prod, Nullable{Function}(prod), Nullable{Function}(prod))
+
+# "Construct a linear operator from a single function."
+# LinearOperator(nrow :: Int, ncol :: Int,
+#                symmetric :: Bool, hermitian :: Bool,
+#                prod :: Function) =
+#   LinearOperator(nrow, ncol, symmetric, hermitian,
+#                  prod, Nullable{Function}(), Nullable{Function}())
+
+
```

Comment:
You can:
```
op = LinearOperator{Float16}(10, 10, true, true, x -> 2*x, x -> 2*x, x -> 2*x)
```
The disadvantage is that you have to rely on the type's default constructor to instantiate an operator built from functions, but no argument is optional. You could omit them like so:
```
op = LinearOperator{Float16}(10, 10, true, true, x -> 2*x, Nullable{Function}(), Nullable{Function}())
```
Still, that's not very convenient. I'd like to find a better design.